# Combining DataFrames With Pandas - Lab

## Introduction

In this lab, you'll gain practice combining DataFrames through concatenation.  You'll also practice executing various types of joins to selectively combine the information stored in the tables!

## Objectives

In this lab you will:

- Use concatenation to combine DataFrames  
- Determine which type of join is preferred for two tables of data and a task  
- Use different types of joins to merge dataframes


## Concatenating DataFrames

Run the cell below to create some sample DataFrames for us to work with.  

In [42]:
import pandas as pd
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                    'B': ['B0', 'B1', 'B2', 'B3'],
                    'C': ['C0', 'C1', 'C2', 'C3'],
                    'D': ['D0', 'D1', 'D2', 'D3']},
                    index=[0, 1, 2, 3])


df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                    'B': ['B4', 'B5', 'B6', 'B7'],
                    'C': ['C4', 'C5', 'C6', 'C7'],
                    'D': ['D4', 'D5', 'D6', 'D7']},
                    index=[4, 5, 6, 7])

df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                    'B': ['B8', 'B9', 'B10', 'B11'],
                    'C': ['C8', 'C9', 'C10', 'C11'], 
                    'D': ['D8', 'D9', 'D10', 'D11']},
                    index=[8, 9, 10, 11])

Now that you have multiple DataFrames to work with, you can execute a concatenation to join them together.  

In the cell below, concatenate the 3 DataFrames together using the appropriate function.   

In [43]:
combined_df = pd.concat([df1, df2, df3])
combined_df

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

**_EXPECTED OUTPUT:_**

<img src="images/er1.png">

## Setting join conditions with concatenation

You can also specify if the concatenation is an **_Outer Join_** or an **_Inner Join_**.  Next, you'll execute an inner join. Before you do, you need to create another table that contains some overlapping index values with a DataFrame that already exists. 

Run the cell below to create the new DataFrame.

In [44]:
df4 = pd.DataFrame({'B': ['B2', 'B3', 'B6', 'B7'],
                    'D': ['D2', 'D3', 'D6', 'D7'],
                    'F': ['F2', 'F3', 'F6', 'F7']},
                    index=[2, 3, 6, 7])

Now, in the cell below, use the `pd.concat()` function to join DataFrames 1 and 4.  However, this time, specify that the `join` is `'inner'`, and `axis=1`. 

In [45]:
# concatenate these 2 DataFrames
# join='inner' means to only return records w/ keys that match btwn the 2 tables (so only index 2 and 3 here)
# axis=1 means concatenate along the columns axis (add the columns from df4 after the columns of df1)
df1_and_4 = pd.concat([df1, df4], join='inner', axis=1)
df1_and_4
# if change axis=1 to axis=0, only returns columns B and D (essentially the matching 'keys') and concatenates
  ### by adding the rows from df4 after the rows from df1, but this makes the new index of the rows all wonky

A   B   C   D   B   D   F
2  A2  B2  C2  D2  B2  D2  F2
3  A3  B3  C3  D3  B3  D3  F3

**_EXPECTED OUTPUT:_**

<img src='images/er2.png'>

You'll notice that in this case, the results contain only the rows with indexes that exist in both tables -- rows 2 and 3.  The resulting table contains the values for each column in both tables for the rows.  

Note that there are many, many ways that you can make full use of the `pd.concat()` function in pandas to join DataFrames together -- these are just a few of the most common examples pulled from the pandas documentation on the subject. For a full view of all the ways you can use `pd.concat()`, see the [pandas documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html)!

## Load data
Now, it's time to move on to working with the Hearthstone cards database.  This database contains information on cards from the popular game, [Hearthstone](https://playhearthstone.com/en-us/)! For full information on the dataset, see the [Kaggle page](https://www.kaggle.com/jeradrose/hearthstone-cards) for this dataset. 

This database consists of the following tables:

* _cards_
* *dust_costs*
* _entourages_
* _mechanics_
* *play_requirements*

Many of rows in each table -- but not all -- correspond to the same cards. As such, each table contains a column called `card_id` which acts as a **_Primary Key_** for each table.  You'll make use of these keys to **_join_** the different tables together into a single DataFrame. You'll also experiment with different types of joins to help us decide exactly what information you wish to combine.  

Simply run the cell below to import the tables from the database as DataFrames.

In [46]:
cards_df = pd.read_csv('cards.csv')
dust_df = pd.read_csv('dust.csv')
entourages_df = pd.read_csv('entourages.csv')
mechanics_df = pd.read_csv('mechanics.csv')
play_requirements_df = pd.read_csv('play_requirements.csv')

Great.  Now, let's set the correct column, `card_id`, as the index column for each of these tables, and then display each to ensure that everything is as expected.  

For each of the DataFrames you created in the cell above, call the `.set_index()` method and pass in `card_id`.  Also set `inplace=True`.  Then, display the `.head()` of each respective DataFrame to ensure everything worked.  

**_NOTE:_** Since you are performing this operation in place, running any cell a second time will result in pandas throwing an error.  If you need to run something a second time, restart the kernel using the jupyter notebook menu at the top of the page.  

In [47]:
cards_df.set_index('card_id', inplace=True)
print(cards_df.shape)
cards_df.head()

(2819, 17)


player_class    type                name      set  \
card_id                                                        
KARA_00_07         MAGE   SPELL       Astral Portal     KARA   
NEW1_008a         DRUID   SPELL   Ancient Teachings  EXPERT1   
BRM_010t2         DRUID  MINION  Druid of the Flame      BRM   
AT_132          NEUTRAL  MINION  Justicar Trueheart      TGT   
OG_141          NEUTRAL  MINION   Faceless Behemoth       OG   

                                                         text  cost  attack  \
card_id                                                                       
KARA_00_07           Summon a random <b>Legendary</b> minion.   1.0     NaN   
NEW1_008a                                        Draw a card.   0.0     NaN   
BRM_010t2                                                 NaN   3.0     2.0   
AT_132      <b>Battlecry:</b> Replace your starting Hero P...   6.0     6.0   
OG_141                                                    NaN  10.0    10.0   

            health     rarity  collectible  \
card_id                                      
KARA_00_07     NaN        NaN          NaN   
NEW1_008a      NaN        NaN          NaN   
BRM_010t2      5.0     COMMON          NaN   
AT_132         3.0  LEGENDARY          1.0   
OG_141        10.0     COMMON          1.0   

                                                       flavor   race  \
card_id                                                                
KARA_00_07                                                NaN    NaN   
NEW1_008a                                                 NaN    NaN   
BRM_010t2                                                 NaN  BEAST   
AT_132      It's like putting racing stripes and a giant s...    NaN   
OG_141      Rejected names: Forty-Foot Faceless, Big ol' N...    NaN   

           how_to_earn how_to_earn_golden targeting_arrow_text faction  \
card_id                                                                  
KARA_00_07         NaN                NaN                  NaN     NaN   
NEW1_008a          NaN                NaN                  NaN     NaN   
BRM_010t2          NaN                NaN                  NaN     NaN   
AT_132             NaN                NaN                  NaN     NaN   
OG_141             NaN                NaN                  NaN     NaN   

            durability  
card_id                 
KARA_00_07         NaN  
NEW1_008a          NaN  
BRM_010t2          NaN  
AT_132             NaN  
OG_141             NaN

In [48]:
dust_df.set_index('card_id', inplace=True)
print(dust_df.shape)
dust_df.head()

(4040, 2)


action  cost
card_id                           
BRM_010t2    CRAFTING_NORMAL    40
BRM_010t2    CRAFTING_GOLDEN   400
BRM_010t2  DISENCHANT_NORMAL     5
BRM_010t2  DISENCHANT_GOLDEN    50
AT_132       CRAFTING_NORMAL  1600

In [49]:
entourages_df.set_index('card_id', inplace=True)
print(entourages_df.shape)
entourages_df.head()

(328, 1)


entourage_card_id
card_id                     
KAR_A10_22        KAR_A10_09
KAR_A10_22        KAR_A10_02
KAR_A10_22        KAR_A10_08
KAR_A10_22        KAR_A10_04
KAR_A10_22        KAR_A10_05

In [50]:
mechanics_df.set_index('card_id', inplace=True)
print(mechanics_df.shape)
mechanics_df.head()

(1079, 1)


mechanic
card_id                      
AT_132              BATTLECRY
GVG_011a  TAG_ONE_TURN_EFFECT
EX1_583             BATTLECRY
LOE_007t            EVIL_GLOW
LOE_007t   ImmuneToSpellpower

In [51]:
play_requirements_df.set_index('card_id', inplace=True)
print(play_requirements_df.shape)
play_requirements_df.head()

(776, 2)


play_requirement  value
card_id                                     
KARA_00_07       REQ_NUM_MINION_SLOTS      1
PRO_001a         REQ_NUM_MINION_SLOTS      1
NAX1_01          REQ_NUM_MINION_SLOTS      1
DS1h_292_H1           REQ_STEADY_SHOT      0
DS1h_292_H1  REQ_MINION_OR_ENEMY_HERO      0

## Executing Joins

Now that you have the tables loaded correctly, we're going to execute some joins. There are four different kinds of joins, which can best be visualized with Venn diagrams:

<img src='images/Image_198_joins.png'>

In these diagrams, each circle represents a DataFrame or SQL Table. The left table is the table you are working with, and the right table is the table you want to join to the table you are working with. You'll start by executing the most common type of join, an **_Inner Join_**.

In the cell below, join `cards_df` with `mechanics_df` using the built-in `.join()` method on the `cards_df` object. 

Pass in the following parameters:
* the table you want to join with, `mechanics_df`
* The `how` parameter set to the type of join you want, `'inner'`

In [62]:
cards_with_mechanics_df = mechanics_df.join(cards_df, how='inner')
# inner is going to return only the records w/ matching keys (card IDs from the right df that match up)
print('left df shape: {}      right df shape: {}      new df shape: {}'.format(mechanics_df.shape, 
                                                                               cards_df.shape,
                                                                               cards_with_mechanics_df.shape))
cards_with_mechanics_df

left df shape: (1079, 1)      right df shape: (2819, 17)      new df shape: (1079, 18)


mechanic player_class    type                    name  \
card_id                                                                    
AT_002               SECRET         MAGE   SPELL                  Effigy   
AT_005t              CHARGE      NEUTRAL  MINION                    Boar   
AT_006              INSPIRE         MAGE  MINION        Dalaran Aspirant   
AT_007            BATTLECRY         MAGE  MINION            Spellslinger   
AT_009          DEATHRATTLE         MAGE  MINION                  Rhonin   
...                     ...          ...     ...                     ...   
XXX_105  ImmuneToSpellpower      NEUTRAL   SPELL        Add 8 to Health.   
XXX_107  ImmuneToSpellpower      NEUTRAL   SPELL         Set Health to 1   
XXX_110           BATTLECRY      NEUTRAL  MINION  Yogg-Saron Test (Auto)   
hexfrog               TAUNT      NEUTRAL  MINION                    Frog   
tt_010               SECRET         MAGE   SPELL             Spellbender   

             set                                               text  cost  \
card_id                                                                     
AT_002       TGT  <b>Secret:</b> When a friendly minion dies, su...   3.0   
AT_005t      TGT                                      <b>Charge</b>   3.0   
AT_006       TGT       <b>Inspire:</b> Gain <b>Spell Damage +1</b>.   4.0   
AT_007       TGT  <b>Battlecry:</b> Add a random spell to each p...   3.0   
AT_009       TGT  <b>Deathrattle:</b> Add 3 copies of Arcane Mis...   8.0   
...          ...                                                ...   ...   
XXX_105    CHEAT  Adds 8 health to a damaged character. Does NOT...   0.0   
XXX_107    CHEAT  Set a character's health to 1, and remove all ...   0.0   
XXX_110    CHEAT  <b>Battlecry:</b> Cast 30 random spells <i>(ta...   0.0   
hexfrog     CORE                                       <b>Taunt</b>   0.0   
tt_010   EXPERT1  <b>Secret:</b> When an enemy casts a spell on ...   3.0   

         attack  health     rarity  collectible  \
card_id                                           
AT_002      NaN     NaN       RARE          1.0   
AT_005t     4.0     2.0        NaN          NaN   
AT_006      3.0     5.0     COMMON          1.0   
AT_007      3.0     4.0     COMMON          1.0   
AT_009      7.0     7.0  LEGENDARY          1.0   
...         ...     ...        ...          ...   
XXX_105     NaN     NaN     COMMON          NaN   
XXX_107     NaN     NaN     COMMON          NaN   
XXX_110     7.0     5.0  LEGENDARY          NaN   
hexfrog     0.0     1.0     COMMON          NaN   
tt_010      NaN     NaN       EPIC          1.0   

                                                    flavor   race how_to_earn  \
card_id                                                                         
AT_002                                  Burning man, brah.    NaN         NaN   
AT_005t                                                NaN  BEAST         NaN   
AT_006    Is he aspiring or inspiring?  Make up your mind!    NaN         NaN   
AT_007   Does he sling spells, or do his spells linger ...    NaN         NaN   
AT_009                              A masterless shamurai.    NaN         NaN   
...                                                    ...    ...         ...   
XXX_105                                                NaN    NaN         NaN   
XXX_107                                                NaN    NaN         NaN   
XXX_110                                                NaN    NaN         NaN   
hexfrog                                                NaN  BEAST         NaN   
tt_010   While it's fun to intercept enemy lightning bo...    NaN         NaN   

        how_to_earn_golden targeting_arrow_text faction  durability  
card_id                                                              
AT_002                 NaN                  NaN     NaN         NaN  
AT_005t                NaN                  NaN     NaN         NaN  
AT_006                 NaN    

Examine the output from the cell above and compare it to the original output of both the `cards_df` and `mechanics_df` DataFrame heads you displayed earlier.  Notice how this now combines the columns from both?

**_Question_**

If you inspect the original `cards_df` DataFrame, you'll notice that it contains  2,819 records.  The result of our inner join, `cards_with_mechanics_df`, contains only 1079 rows.  Why?

Write your answer below this line:
________________________________________________________________________________
The mechanics_df had 1079 keys so there could only be 1079 possible key matches (and thus 1079 possible rows) to concatenate from the cards_df along the columns axis.

## Other Types of Joins

By default, the `.join()` method performs a left join if no parameter is passed in for `how=`.  In the cell below, perform a **_Left Join_** of `cards_with_mechanics_df` and `play_requirements_df`, with `cards_with_mechanics_df` as the left table.  

Then, display `left_join_df` to inspect our results. 

In [72]:
# performs a left join as a default if how= is not specified
# a left join returns all the records from the left table plus any records from the right table that have
## a key that matches any key from the records in the left table (in this case the key is the index, so the card ID)
left_join_df = cards_with_mechanics_df.join(play_requirements_df)
print('left df shape: {}      right df shape: {}      new df shape: {}'.format(cards_with_mechanics_df.shape, 
                                                                               play_requirements_df.shape,
                                                                               left_join_df.shape))
print(left_join_df.info())
left_join_df

left df shape: (1079, 18)      right df shape: (776, 2)      new df shape: (1226, 20)
<class 'pandas.core.frame.DataFrame'>
Index: 1226 entries, AT_002 to tt_010
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   mechanic              1226 non-null   object 
 1   player_class          1226 non-null   object 
 2   type                  1226 non-null   object 
 3   name                  1226 non-null   object 
 4   set                   1226 non-null   object 
 5   text                  1189 non-null   object 
 6   cost                  1157 non-null   float64
 7   attack                968 non-null    float64
 8   health                952 non-null    float64
 9   rarity                886 non-null    object 
 10  collectible           825 non-null    float64
 11  flavor                825 non-null    object 
 12  race                  215 non-null    object 
 13  how_to_earn           143 non-null 

mechanic player_class    type                    name  \
card_id                                                                    
AT_002               SECRET         MAGE   SPELL                  Effigy   
AT_005t              CHARGE      NEUTRAL  MINION                    Boar   
AT_006              INSPIRE         MAGE  MINION        Dalaran Aspirant   
AT_007            BATTLECRY         MAGE  MINION            Spellslinger   
AT_009          DEATHRATTLE         MAGE  MINION                  Rhonin   
...                     ...          ...     ...                     ...   
XXX_105  ImmuneToSpellpower      NEUTRAL   SPELL        Add 8 to Health.   
XXX_107  ImmuneToSpellpower      NEUTRAL   SPELL         Set Health to 1   
XXX_110           BATTLECRY      NEUTRAL  MINION  Yogg-Saron Test (Auto)   
hexfrog               TAUNT      NEUTRAL  MINION                    Frog   
tt_010               SECRET         MAGE   SPELL             Spellbender   

             set                                               text  cost  \
card_id                                                                     
AT_002       TGT  <b>Secret:</b> When a friendly minion dies, su...   3.0   
AT_005t      TGT                                      <b>Charge</b>   3.0   
AT_006       TGT       <b>Inspire:</b> Gain <b>Spell Damage +1</b>.   4.0   
AT_007       TGT  <b>Battlecry:</b> Add a random spell to each p...   3.0   
AT_009       TGT  <b>Deathrattle:</b> Add 3 copies of Arcane Mis...   8.0   
...          ...                                                ...   ...   
XXX_105    CHEAT  Adds 8 health to a damaged character. Does NOT...   0.0   
XXX_107    CHEAT  Set a character's health to 1, and remove all ...   0.0   
XXX_110    CHEAT  <b>Battlecry:</b> Cast 30 random spells <i>(ta...   0.0   
hexfrog     CORE                                       <b>Taunt</b>   0.0   
tt_010   EXPERT1  <b>Secret:</b> When an enemy casts a spell on ...   3.0   

         attack  health     rarity  collectible  \
card_id                                           
AT_002      NaN     NaN       RARE          1.0   
AT_005t     4.0     2.0        NaN          NaN   
AT_006      3.0     5.0     COMMON          1.0   
AT_007      3.0     4.0     COMMON          1.0   
AT_009      7.0     7.0  LEGENDARY          1.0   
...         ...     ...        ...          ...   
XXX_105     NaN     NaN     COMMON          NaN   
XXX_107     NaN     NaN     COMMON          NaN   
XXX_110     7.0     5.0  LEGENDARY          NaN   
hexfrog     0.0     1.0     COMMON          NaN   
tt_010      NaN     NaN       EPIC          1.0   

                                                    flavor   race how_to_earn  \
card_id                                                                         
AT_002                                  Burning man, brah.    NaN         NaN   
AT_005t                                                NaN  BEAST         NaN   
AT_006    Is he aspiring or inspiring?  Make up your mind!    NaN         NaN   
AT_007   Does he sling spells, or do his spells linger ...    NaN         NaN   
AT_009                              A masterless shamurai.    NaN         NaN   
...                                                    ...    ...         ...   
XXX_105                                                NaN    NaN         NaN   
XXX_107                                                NaN    NaN         NaN   
XXX_110                                                NaN    NaN         NaN   
hexfrog                                                NaN  BEAST         NaN   
tt_010   While it's fun to intercept enemy lightning bo...    NaN         NaN   

        how_to_earn_golden targeting_arrow_text faction  durability  \
card_id                                                               
AT_002                 NaN                  NaN     NaN         NaN   
AT_005t                NaN                  NaN     NaN         NaN   
AT_006                 NaN

Note that the results of this sort of join are dependent upon the position of each table--if you were to make `cards_with_mechanics_df` the right table and `play_requirements_df` the left table and then perform a **_Right Join_**, our results would be the same. 

**_Question:_**

Describe what was included from each table in this join.

Write your answer below this line:
________________________________________________________________________________
This join included all records (rows and cols) from the left table (cards_with_mechanics_df). To this it tacked on to the right side of all those cols the 2 cols 'play_requirement' and 'value' that were the only 2 cols of the play_requirement_df. However it only did this for keys that matched those already found in the left df.

#### Outer Joins

In the cell below, perform an outer join between `cards_df` and `dust_df`. Since these tables contain columns with the same name, we'll need to specify a suffix for at least one of them, so that the column can be renamed to avoid a naming collision. 

During this join, set the `rsuffix` parameter to `_dust`

In [74]:
# outer join returns all records from both DataFrames
## need to specify a suffix to add to the keys for one of the two DataFrames since they have matching keys
### that would otherwise cause problems
outer_join_df = cards_df.join(dust_df, rsuffix='_dust', how='outer')
print('left df shape: {}      right df shape: {}      new df shape: {}'.format(cards_df.shape, 
                                                                               dust_df.shape,
                                                                               outer_join_df.shape))
print(outer_join_df.info())
outer_join_df

left df shape: (2819, 17)      right df shape: (4040, 2)      new df shape: (5849, 19)
<class 'pandas.core.frame.DataFrame'>
Index: 5849 entries, AT_001 to tt_010a
Data columns (total 19 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   player_class          5846 non-null   object 
 1   type                  5848 non-null   object 
 2   name                  5848 non-null   object 
 3   set                   5848 non-null   object 
 4   text                  5284 non-null   object 
 5   cost                  5198 non-null   float64
 6   attack                3568 non-null   float64
 7   health                3597 non-null   float64
 8   rarity                4429 non-null   object 
 9   collectible           3821 non-null   float64
 10  flavor                3807 non-null   object 
 11  race                  1006 non-null   object 
 12  how_to_earn           704 non-null    object 
 13  how_to_earn_golden    749 non-nul

player_class    type         name      set  \
card_id                                              
AT_001          MAGE   SPELL  Flame Lance      TGT   
AT_001          MAGE   SPELL  Flame Lance      TGT   
AT_001          MAGE   SPELL  Flame Lance      TGT   
AT_001          MAGE   SPELL  Flame Lance      TGT   
AT_002          MAGE   SPELL       Effigy      TGT   
...              ...     ...          ...      ...   
tt_010          MAGE   SPELL  Spellbender  EXPERT1   
tt_010a         MAGE  MINION  Spellbender  EXPERT1   
tt_010a         MAGE  MINION  Spellbender  EXPERT1   
tt_010a         MAGE  MINION  Spellbender  EXPERT1   
tt_010a         MAGE  MINION  Spellbender  EXPERT1   

                                                      text  cost  attack  \
card_id                                                                    
AT_001                         Deal $8 damage to a minion.   5.0     NaN   
AT_001                         Deal $8 damage to a minion.   5.0     NaN   
AT_001                         Deal $8 damage to a minion.   5.0     NaN   
AT_001                         Deal $8 damage to a minion.   5.0     NaN   
AT_002   <b>Secret:</b> When a friendly minion dies, su...   3.0     NaN   
...                                                    ...   ...     ...   
tt_010   <b>Secret:</b> When an enemy casts a spell on ...   3.0     NaN   
tt_010a                                                NaN   1.0     1.0   
tt_010a                                                NaN   1.0     1.0   
tt_010a                                                NaN   1.0     1.0   
tt_010a                                                NaN   1.0     1.0   

         health  rarity  collectible  \
card_id                                
AT_001      NaN  COMMON          1.0   
AT_001      NaN  COMMON          1.0   
AT_001      NaN  COMMON          1.0   
AT_001      NaN  COMMON          1.0   
AT_002      NaN    RARE          1.0   
...         ...     ...          ...   
tt_010      NaN    EPIC          1.0   
tt_010a     3.0    EPIC          NaN   
tt_010a     3.0    EPIC          NaN   
tt_010a     3.0    EPIC          NaN   
tt_010a     3.0    EPIC          NaN   

                                                    flavor race how_to_earn  \
card_id                                                                       
AT_001   It's on the rack next to ice lance, acid lance...  NaN         NaN   
AT_001   It's on the rack next to ice lance, acid lance...  NaN         NaN   
AT_001   It's on the rack next to ice lance, acid lance...  NaN         NaN   
AT_001   It's on the rack next to ice lance, acid lance...  NaN         NaN   
AT_002                                  Burning man, brah.  NaN         NaN   
...                                                    ...  ...         ...   
tt_010   While it's fun to intercept enemy lightning bo...  NaN         NaN   
tt_010a                                                NaN  NaN         NaN   
tt_010a                                                NaN  NaN         NaN   
tt_010a                                                NaN  NaN         NaN   
tt_010a                                                NaN  NaN         NaN   

        how_to_earn_golden targeting_arrow_text faction  durability  \
card_id                                                               
AT_001                 NaN                  NaN     NaN         NaN   
AT_001                 NaN                  NaN     NaN         NaN   
AT_001                 NaN                  NaN     NaN         NaN   
AT_001                 NaN                  NaN     NaN         NaN   
AT_002                 NaN                  NaN     NaN         NaN   
...                    ...                  ...     ...         ...   
tt_010                 NaN                  NaN     NaN         NaN   
tt_010a                NaN                  NaN     NaN         NaN   
tt_010a                NaN                  NaN     NaN  

Inspect the output above.  Note that the naming collision has been avoided by renaming the `cost` column from the right table to `cost_dust`.  

## Summary

In this lab, you learned how to:

* Concatenate multiple DataFrames together into a single DataFrame
* Understand and execute the various types of joins (inner, outer, left, and right joins)